### PR TITLE
improvement: read old IRs + raise warnings

### PR DESCRIPTION
**Pull request by alexjuda**
# The problem

_Please describe the motivation for this PR. Bonus points for code examples, error messages, and screenshots._

# This PR's solution

Blocked on ~https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/107~, ~https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/108~.

1. Make `TaskDef.output_metadata` optional. This allows deserializing IRs from `orquestra-sdk<=v0.45.1`
2. Add `_compat` module for a single place where we handle cases where the task def is `None`.
3. Add regression test in form of a workflow def IR serialized from `orquestra-sdk==0.45.1` and `0.44.0`.
4. Add validator on `ir.WorkflowDef` that raises warnings when deserializing an IR generated with a different version of the SDK.

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`

### CODE REVIEW

File: src/orquestra/sdk/schema/ir.py [hunk 1]
```diff
@@ -414,3 +417,41 @@ class WorkflowDef(BaseModel):
...
+                        "different version of Orquestra Workflow SDK. "
+                        "Please consider re-running your workflow or installing "
+                        f"'orquestra-sdk=={v.sdk_version.original}'"
+                    ),
+                    actual=current_version,
+                    needed=v.sdk_version,
+                )
+            )
+
+        return v
```
Comment (jamesclark-Zapata) [inline src/orquestra/sdk/schema/ir.py:457]: **suggestion:** ‏I think this makes sense to be a semver check vs an entire version check.
Right now, because we're pre-1.0.0 it would give the same result (all pre v1 minor versions are technically semver breaking changes). 
My reasoninfg is that this would give us a clear line for "IR breakages are a major version bump" and would mean we don't bother people with warnings a lot between v1 and v2.

File: src/orquestra/sdk/schema/ir.py [hunk 2]
```diff
@@ -414,3 +417,41 @@ class WorkflowDef(BaseModel):
...
+            return v
+
+        if v.sdk_version != current_version:
+            warnings.warn(
+                exceptions.VersionMismatch(
+                    (
+                        "Attempting to read a workflow definition generated with a "
+                        "different version of Orquestra Workflow SDK. "
+                        "Please consider re-running your workflow or installing "
+                        f"'orquestra-sdk=={v.sdk_version.original}'"
```
Comment (jamesclark-Zapata) [inline src/orquestra/sdk/schema/ir.py:450]: **suggestion:** ‏This is likely to appear a lot, at least in the short term. Could you add a new page to the docs and link it here that says something like:
* Explain *why* this is an issue
* Say that things might be OK and possible scenarios that will fail

File: tests/sdk/v2/conversions/test_yaml_exporter.py [hunk 3]
```diff
@@ -18,950 +18,967 @@ def _model_dict(model_obj):
     return json.loads(model_obj.json(exclude_none=True))
```
Comment (jamesclark-Zapata) [inline tests/sdk/v2/conversions/test_yaml_exporter.py:18]: the changes here seem unrelated to the main focus of the PR (at least I'm finding it hard to see a link).
wdyt about making this a separate commit/PR after it lands in `breaking-changes`?

File: tests/sdk/v2/schema/test_compat.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+SNAPSHOT_VERSIONS = [
+    # Last version without ``ir.WorkflowDef.metadata``
+    "0.44.0",
+    # Last version without ``ir.TaskDef.output_metadata``
+    "0.45.1",
+]
+
+
+class TestNOutputs:
+    class TestModernIR:
```
Comment (jamesclark-Zapata) [inline tests/sdk/v2/schema/test_compat.py:49]: **nitpick:** ‏What is "modern IR"? "Current IR" seems to be more accurate

File: tests/sdk/v2/schema/test_compat.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            n_outputs = _compat.n_outputs(task_def, task_inv)
+
+            # Then
+            assert n_outputs == 1
+
+        # No test for "multi_output" because "n_outputs" isn't reliable with old IRs.
+
+
+class TestResultIsPacked:
+    class TestModernIR:
```
Comment (jamesclark-Zapata) [inline tests/sdk/v2/schema/test_compat.py:109]: same here

File: tests/sdk/v2/schema/test_ir.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    class TestRegression:
+        """
+        Validates we're able to deserialize workflow defs generated with old versions
+        of the SDK. We should be careful about the assertions; probably we should never
+        change them to test for regressions.
+        """
+
+        @staticmethod
+        def test_new_ir():
```
Comment (jamesclark-Zapata) [inline tests/sdk/v2/schema/test_ir.py:45]: same here "new_ir" vs "current_ir"

**Review by jamesclark-Zapata — COMMENTED** [review]
looks good, I have a few suggestions.

Is there a way to test this out with either a runtime (CE) or public API using a runtime (again, CE).

File: tests/sdk/v2/conversions/test_yaml_exporter.py (hunk 3 above)
Comment (alexjuda) [reply tests/sdk/v2/conversions/test_yaml_exporter.py:18]: The changes in this file are caused by a combination of factors:
1. I've made `WorkflowDef` raise a warning when the dumped SDK version is different than the parse-time SDK version.
2. We treat warnings as errors in our test suite.
3. Pytest lets silencing warnings on a per test-case basis, but not on per-statement. [Pytest docs](https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html) mention `with warnings.catch_warnings(): warnings.simplefilter("ignore")` but I couldn't make it work with our stuff. Probably (2) captures warnings as errors regardless of silencing them.
4. Workflow defs in this file were hardcoded as module constants evaluated at module load time instead of fixtures evaluated at test execution time. I turned them into thunks.

I can see how these changes make this PR difficult to read. I'll be happy to extract these changes into a separate PR.

**Comment by alexjuda** [comment]
@jamesclark-Zapata 

> Is there a way to test this out with either a runtime (CE) or public API using a runtime (again, CE).

Are you thinking about coarse-grained integration tests between the Python API and CE response mocks or sth else?

**Comment by jamesclark-Zapata** [comment]
> @jamesclark-Zapata
> 
> > Is there a way to test this out with either a runtime (CE) or public API using a runtime (again, CE).
> 
> Are you thinking about coarse-grained integration tests between the Python API and CE response mocks or sth else?

Yeah, mock out the API responses (based on submitting workflows in those versions) and try and e2e test them

File: src/orquestra/sdk/schema/ir.py (hunk 1 above)
Comment (jamesclark-Zapata) [reply src/orquestra/sdk/schema/ir.py:457]: It shouldn't take long and we'd just line ourselves up for tech debt.

```
if (current_version.major > 0 and current_version.major != v.sdk_version.major) or (current_version.major == 0 and current_version != v.sdk_version):
    pass
```

File: tests/sdk/v2/conversions/test_yaml_exporter.py (hunk 3 above)
Comment (alexjuda) [reply tests/sdk/v2/conversions/test_yaml_exporter.py:18]: Split into https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/107

File: src/orquestra/sdk/schema/ir.py (hunk 1 above)
Comment (alexjuda) [reply src/orquestra/sdk/schema/ir.py]: Added in [d106280](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/105/commits/d10628093c867f02c2a724c3359a246e3fc5aa7f). WDYT?

File: tests/sdk/v2/schema/test_compat.py (hunk 4 above)
Comment (alexjuda) [reply tests/sdk/v2/schema/test_compat.py]: Fixed in 5b9a0abce3

File: tests/sdk/v2/schema/test_compat.py (hunk 5 above)
Comment (alexjuda) [reply tests/sdk/v2/schema/test_compat.py]: Fixed in 5b9a0abce3

**Comment by alexjuda** [comment]
@jamesclark-Zapata

> Yeah, mock out the API responses (based on submitting workflows in those versions) and try and e2e test them

Added as a separate ticket: https://zapatacomputing.atlassian.net/browse/ORQSDK-815

File: src/orquestra/sdk/schema/ir.py (hunk 2 above)
Comment (alexjuda) [reply src/orquestra/sdk/schema/ir.py]: First draft added in 7903a35008d. LMK WYT.

I hope I correctly predicted the URL slugs, haha. I based this on the links I've seen in my local builds but I guess we'll see when the docs are rebuilt.

File: docs/guides/version-compatibility.rst [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+======================
+
+When you're running workflows remotely, you have a local version of Orquestra Workflow SDK on your local machine or the Jupyter environment in Orquestra Portal.
+An Intermediate Representation (IR) is generated from your ``@workflow``-decorated function and sent over the wire to the remote runtime.
+
+The remote runtime also comes with its own installation of the Workflow SDK to interpret the workflow IR, execute tasks, and store results.
+The Workflow SDK version installed on the server side is selected to match the client's one.
+
+Finally, you're likely to interact with the workflow run to query its status, peek into the graph of task runs, or access the data they produced.
+This interaction can happen a long time after the workflow was originally executed; the workflow data lives longer than the Workflow SDK code that produced it.
```
Comment (jamesclark-Zapata) [inline docs/guides/version-compatibility.rst:20]: I tried to epxand on some points, idk if it worked.

What I was trying to get across is that this version coupling doesn't only exist for remote workflows, it exists locally too. The only special behaviour is the matching of the SDK version.

```suggestion
Workflow Version Coupling
=========================

When you write a workflow, the Orquestra Workflow SDK generates an Intermediate Representation (IR) from your ``@workflow``-decorated function. When you "run" the workflow, this IR is sent to a runtime for execution. Inside the IR, we store the version of the Workflow SDK that generated it.

When executing the workflow remotely, the remote runtime reads the Workflow SDK version from the IR and uses the same version to interpret the IR, execute tasks, and store results.

Once a workflow run has been started, you will query its status, peek into its graph of task runs, or access the data it produced.

Sometimes these interactions can happen a long time after the workflow was originally executed; the workflow data lives longer than the Workflow SDK version that produced it.

In the meantime, the internals of the Workflow SDK may have changed resulting in unexpected behavior.
```

Comment (alexjuda) [reply docs/guides/version-compatibility.rst]: Yeah, makes more sense to me. Added in 5695438b85

**Review by jamesclark-Zapata — APPROVED** [review]
